app/code/Magento/Inventory/Model/ReservationBuilder.php

```diff
@@ -0,0 +1,181 @@
+     * @var int
+     */
+    private $stockId;
+
+    /**
+     * @var string
+     */
+    private $sku;
+
+    /**
+     * @var float
+     */
+    private $quantity;
+
+    /**
+     * @var string
+     */
+    private $metadata;
+
+    /**
+     * Object Manager instance
+     *
+     * @var \Magento\Framework\ObjectManagerInterface
+     */
+    private $objectManager = null;
+
+    /**
+     * @var \Magento\Inventory\Model\ReservationBuilder\Validator\ReservationBuilderValidatorInterface
+     */
+    private $reservationBuilderValidator;
+
+    /**
+     * String Service instance
+     *
+     * @var \Magento\Inventory\Model\String\Service
+     */
+    private $stringService;
+
+    /**
+     * ReservationBuilder constructor.
+     *
+     * @param \Magento\Framework\ObjectManagerInterface $objectManager
+     */
+    public function __construct(
+        \Magento\Framework\ObjectManagerInterface $objectManager,
+        \Magento\Inventory\Model\ReservationBuilder\Validator\ReservationBuilderValidatorInterface $reservationBuilderValidator,
+        \Magento\Inventory\Model\String\Service $stringService
+    ) {
+        $this->objectManager = $objectManager;
+        $this->reservationBuilderValidator = $reservationBuilderValidator;
+        $this->stringService = $stringService;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function setReservationId($reservationId): ReservationBuilder
+    {
+        $this->reservationId = $reservationId;
+        return $this;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function setStockId(int $stockId): ReservationBuilder
+    {
+        $this->stockId = $stockId;
+        return $this;
+    }
+
+    /**
+     * @return int|null
+     */
+    public function getStockId()
+    {
+        return $this->stockId;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function setSku(string $sku): ReservationBuilder
+    {
+        $this->sku = $sku;
+        return $this;
+    }
+
+    /**
+     * @return string|null
+     */
+    public function getSku()
+    {
+        return $this->sku;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function setQuantity(float $quantity): ReservationBuilder
+    {
+        $this->quantity = $quantity;
+        return $this;
+    }
+
+    /**
+     * @return float|null
+     */
+    public function getQuantity()
+    {
+        return $this->quantity;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function setMetadata($metadata): ReservationBuilder
+    {
+        $this->metadata = $metadata;
+        return $this;
+    }
+
+    /**
+     * @inheritdoc
+     */
+    public function build(): ReservationInterface
+    {
+        $validationResult = $this->reservationBuilderValidator->validate($this);
+
+        if (!$validationResult->isValid()) {
+            throw new ValidationException($validationResult);
+        }
+
+        $data = [
+            ReservationInterface::RESERVATION_ID => $this->reservationId,
```

Comment:
no need to pass RESERVATION_ID

moreover not sure that Builder should provide a corresponding method `setReservationId` if we rely on auto generated IDs

Reply:
>no need to pass RESERVATION_ID

if we don't allow the ReservationBuilder to be able to instantiate a Reservation object with an ID this means that the ReservationBuilder can't be used to create a representation of persisted entities.

I'm not saying this is wrong, just asking for a confirmation.